### PR TITLE
Make etcd cache size configurable

### PR DESCRIPTION
**Pull request by jsravn**
Instead of the prior 50K limit, allow users to specify a more sensible size for their cluster.

I'm not sure what a sensible default is here. I'm still experimenting on my own clusters. 50 gives me a 270MB max footprint. 50K caused my apiserver to run out of memory as it exceeded >2GB. I believe that number is far too large for most people's use cases.

There are some other fundamental issues that I'm not addressing here:
- Old etcd items are cached and potentially never removed (it stores using modifiedIndex, and doesn't remove the old object when it gets updated)
- Cache isn't LRU, so there's no guarantee the cache remains hot. This makes its performance difficult to predict. More of an issue with a smaller cache size.
- 1.2 etcd entries seem to have a larger memory footprint (I never had an issue in 1.1, even though this cache existed there). I suspect that's due to image lists on the node status.

This is provided as a fix for #23323

### CODE REVIEW

**Comment by jsravn** [comment]
I think that this change requires making the cache LRU, actually. Having a small randomly evicting cache is almost useless imho...

File: pkg/storage/etcd/etcdtest/etcdtest.go [hunk 1]
```diff
@@ -21,6 +21,8 @@ import (
 	"path"
 )
 
+const CacheSize int = 50
```
Comment (lavalamp) [inline pkg/storage/etcd/etcdtest/etcdtest.go]: Comment.

**Comment by lavalamp** [comment]
I think converting this to an LRU cache, and basing the size off of the current knob for the watch window (instead of introducing an additional knob) maybe be more useful than this. Does this change actually affect apiserver's behavior? I would expect a bigger cache size to use more memory without really giving any benefit?

**Comment by lavalamp** [comment]
Ah, this is the cache we use to save time deserializing. So yes, this needs to eventually be made into an LRU cache. But actually small + randomly evicting is not that bad for this cache.

File: cmd/kube-apiserver/app/options/options.go [hunk 2]
```diff
@@ -230,6 +231,7 @@ func (s *APIServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.EtcdConfig.CertFile, "etcd-certfile", s.EtcdConfig.CertFile, "SSL certification file used to secure etcd communication")
 	fs.StringVar(&s.EtcdConfig.CAFile, "etcd-cafile", s.EtcdConfig.CAFile, "SSL Certificate Authority file used to secure etcd communication")
 	fs.BoolVar(&s.EtcdConfig.Quorum, "etcd-quorum-read", s.EtcdConfig.Quorum, "If true, enable quorum read")
+	fs.IntVar(&s.EtcdConfig.CacheSize, "etcd-cache-size", s.EtcdConfig.CacheSize, "Number of etcd entries to cache in memory")
```
Comment (lavalamp) [inline cmd/kube-apiserver/app/options/options.go]: This cache is actually for deserialized objects. It's not got much to do with etcd. "deserialization-cache-size"?

**Comment by lavalamp** [comment]
OK, I'm actually pretty OK with this change if you can make the flag name/description more accurate.

File: cmd/kube-apiserver/app/options/options.go (hunk 2 above)
Comment (jsravn) [inline cmd/kube-apiserver/app/options/options.go]: Ok, I see your point. Maybe 'resource-cache-size' or 'storage-cache-size' as other options.

**Comment by jsravn** [comment]
> Ah, this is the cache we use to save time deserializing. So yes, this needs to eventually be made into an LRU cache. But actually small + randomly evicting is not that bad for this cache.

Yeah the LRU cache can be done in another PR. I'm curious why you think it won't be so bad as-is, though.

Also, what do you think about the default value I've used? 150 vs the original of 50k.

**Comment by lavalamp** [comment]
> Yeah the LRU cache can be done in another PR. I'm curious why you think it won't be so bad as-is, though.

I said that just by reasoning about the behavior--often used things will get quickly replaced if evicted but rarely used things won't--but the first hit on a search seems to confirm: http://danluu.com/2choices-eviction/ Random looks even better than I expected.

> Also, what do you think about the default value I've used? 150 vs the original of 50k.

Without doing any actual reasoning, my intuition is to go for something a bit bigger than that, maybe something in the range 1k - 5k?

We should really count the number of hits/misses this cache gets and export it as a Prometheus metric, then we'd have actual numbers to go off of. My guess is each entry gets like 2-7 hits and that's it?

File: pkg/genericapiserver/genericapiserver.go [hunk 3]
```diff
@@ -59,8 +59,9 @@ import (
 )
 
 const (
-	DefaultEtcdPathPrefix = "/registry"
-	globalTimeout         = time.Minute
+	DefaultEtcdPathPrefix           = "/registry"
+	DefaultDeserializationCacheSize = 150
```
Comment (lavalamp) [inline pkg/genericapiserver/genericapiserver.go]: Maybe let's start at 1k and see if that breaks our performance tests?

**Comment by jsravn** [comment]
>  http://danluu.com/2choices-eviction/ Random looks even better than I expected.

It also shows how random eviction performs worse the smaller the cache (although he doesn't show smaller than a 64K cache  - I suspect the miss rate starts to drop off exponentially as cache size decreases beyond that). I don't know what workload he used, but I suspect it also has a more uniform chance of data access. Whereas in our case we have etcd objects getting frequently updated, so that the majority are old objects which will never be used again. Given large enough cache size though it should behave okay, just inefficiently.

**Comment by lavalamp** [comment]
LGTM- please squash!

**Comment by jsravn** [comment]
@lavalamp squashed

File: pkg/genericapiserver/genericapiserver.go [hunk 4]
```diff
@@ -59,8 +59,9 @@ import (
 )
 
 const (
-	DefaultEtcdPathPrefix = "/registry"
-	globalTimeout         = time.Minute
+	DefaultEtcdPathPrefix           = "/registry"
+	DefaultDeserializationCacheSize = 1000
```
Comment (wojtek-t) [inline pkg/genericapiserver/genericapiserver.go]: Can you please leave default as it was before (i.e. 50.000)?

**Comment by jsravn** [comment]
@wojtek-t Ok, I changed it back to 50k.

But am not happy about it :). In my cluster, that will result in a ~10GB heap when fully populated. With 99% old useless stuff. At least I can configure it now.

I hope you will consider opening a new issue to investigate replacing it with a more efficient, smaller LRU cache. Or point me to the load tests so I can have a look.

**Comment by wojtek-t** [comment]
@jsravn - what we should do is to set based on the initial cluster size in our salt scripts (similarly to how we e.g. set limits for heapster) That way, default value can be small for small clusters and big for big clusters.

**Comment by wojtek-t** [comment]
BTW - I hope that as soon as we migrate to using protobufs, we would be able to get rid of this cache completely. [So in 1.3 timeframe hopefully]

**Comment by lavalamp** [comment]
@wojtekt Was the 50k a result of measuring something?

**Comment by wojtek-t** [comment]
@lavalamp - it was (although it was more than half an year ago, and I don't remember what exactly experiments I was doing :))

**Comment by lavalamp** [comment]
LGTM

**Comment by ncdc** [comment]
@lavalamp I'd like to suggest this for 1.2.x, WDYT?

**Comment by lavalamp** [comment]
@ncdc why?

**Comment by ncdc** [comment]
To allow customers to avoid running out of memory in their apiservers

**Comment by lavalamp** [comment]
Is this a problem lots of people are experiencing? I'm in favor of only
cherrypicking absolutely essential changes. Anyway, you can add the
cherrypick-candidate label if you want, and see what happens :)

**Comment by ncdc** [comment]
@lavalamp never mind on the cherry-pick unless others really want it (OpenShift will pull it into our vendored copy of Kube)

**Comment by hongchaodeng** [comment]
@lavalamp 
I found that federatd/ also has an API server and it doesn't have any change from this. Specifically, it doesn't has default cache size.
Should we add it?

**Comment by wojtek-t** [comment]
@hongchaodeng - at least not now in my opinion. To be honest, I would like to get rid of this cache completely. I hope that as soon as we have protobufs this should be possible.

**Pull request merged**